x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/ssl/SSLReloadDuringStartupIntegTests.java

```diff
@@ -0,0 +1,130 @@
+        Path origKeystorePath = getDataPath("/org/elasticsearch/xpack/security/transport/ssl/certs/simple/testnode.jks");
+        Path nodeSpecificPath = tmpEnv.configFile().resolve("testnode.jks");
+        try {
+            Files.copy(origKeystorePath, nodeSpecificPath, StandardCopyOption.REPLACE_EXISTING);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        Settings.Builder builder = Settings.builder()
+            .put(settings.filter((s) -> s.startsWith("xpack.security.transport.ssl.") == false), false);
+        MockSecureSettings secureSettings = new MockSecureSettings();
+        builder.setSecureSettings(secureSettings);
+        builder.put("xpack.security.transport.ssl.keystore.path", nodeSpecificPath)
+            .put("resource.reload.interval.high", "1s")
+            .put("xpack.security.transport.ssl.enabled", true);
+        secureSettings.setString("xpack.security.transport.ssl.keystore.secure_password", "testnode");
+        return builder.build();
+    }
+
+    @Override
+    protected boolean transportSSLEnabled() {
+        return true;
+    }
+
+    public void testReloadDuringStartup() throws Exception {
+        final String node = randomFrom(internalCluster().getNodeNames());
+        final Environment env = internalCluster().getInstance(Environment.class, node);
+        final CountDownLatch latch = new CountDownLatch(2);
+        final CountDownLatch writtenLatch = new CountDownLatch(1);
+
+        // restart the node
+        internalCluster().restartNode(node, new RestartCallback() {
+            @Override
+            public Settings onNodeStopped(String nodeName) throws Exception {
+                Path origKeystorePath = getDataPath("/org/elasticsearch/xpack/security/transport/ssl/certs/simple/testnode.jks");
+                Path keystorePath = env.configFile().resolve("testnode.jks");
+                Path updatedKeystorePath = getDataPath("/org/elasticsearch/xpack/security/transport/ssl/certs/simple/testnode_updated.jks");
+                assertTrue(Files.exists(keystorePath));
+                // replace the keystore
+                copyAndAtomicMoveIfPossible(updatedKeystorePath, keystorePath);
+                new Thread(() -> {
+                    latch.countDown();
+                    try {
+                        latch.await();
+                        Thread.sleep(randomLongBetween(1L, 2000L));
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                    }
+
+                    try {
+                        copyAndAtomicMoveIfPossible(origKeystorePath, keystorePath);
+                        writtenLatch.countDown();
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                }).start();
+
+                latch.countDown();
```

Comment:
`latch` is a way to allow for the test to synchronize between a thread that performs the file modification during startup to be constructed and execution of node restart to continue.

`writtenLatch` provides a way for the test itself to wait until the file has been written; technically it is not necessary but avoids us trying to execute the checks on cluster size and connections until we know that connections are actually expected to succeed.

A phaser could be used here but we know upfront the number of registrants and do not need to be used repeatedly for waiting, so latches suffice for this test.

I'm happy to elaborate more if necessary.

Reply:
I also added comments in the test https://github.com/elastic/elasticsearch/pull/54999/commits/2db3a9a6b724bdf6adf01972a9be8e2504a74698

Reply:
Thanks for the explanation. Maybe I didn't ask the right question. I understand what the 2 `latch` trying to do. I just wasn't sure whether they are necessary. Since the original failure is a rare race condition which is hard to re-produce. So would it be better if we just let the threads run without any synchronisation for a possibly more real simulation?

Reply:
I think the synchronization helps get closer to the possibility of that rare condition by ensuring one thread doesn't just jump ahead which would mean that the two actions do not happen concurrently.